python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,601 @@
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import datetime
+import sys
+import uuid
+from decimal import ROUND_HALF_UP, Decimal
+from functools import singledispatch
+from typing import Any
+
+from iceberg.types import PrimitiveType
+
+if sys.version_info >= (3, 8):
+    from functools import singledispatchmethod  # pragma: no cover
+else:
+    from singledispatch import singledispatchmethod  # pragma: no cover
+
+import pytz
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    Singleton,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+
+"""
+Iceberg literal is wrapper class used in expressions, which return unbound predicates
+It's being organized as below
+Literal
+|-- AboveMax
+|-- BelowMin
+|-- BaseLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
+    |-- ComparableLiteral
+        |-- BooleanLiteral
+        |-- IntegerLiteral
+        |-- LongLiteral
+        |-- FloatLiteral
+        |-- DoubleLiteral
+        |-- DateLiteral
+        |-- StringLiteral
+        |-- TimeLiteral
+        |-- TimestampLiteral
+        |-- DecimalLiteral
+        |-- UUIDLiteral
+"""
+
+
+class Literal:
```

Comment:
I think that this should be parameterized so you know what kind of Python object is returned by `literal.value`.

I think that should work with `@singledispatch` because it can use `def literal(value: T) -> BaseLiteral[T]:`